pip/req/req_set.py

```diff
@@ -432,8 +458,18 @@ def prepare_files(self, finder):
                         )
 
                 if not self.ignore_dependencies:
-                    for subreq in dist.requires(
+
+                    for missing in RequirementSet._missing_extras(
+                            dist.extras,
                             req_to_install.extras):
+                        logger.warning(
+                            '%s does not provide the extra \'%s\'',
+                            str(dist), missing)
+
+                    for subreq in dist.requires(
+                            RequirementSet._available_extras(
+                                dist.extras,
+                                req_to_install.extras)):
```

Comment:
The static methods seem a bit much for basically doing

    set(req_to_install.extras) & set(dist.extras)

and

        set(req_to_install.extras) - set(dist.extras)